### DISCUSSION TITLE
Merging(concatenating) items from multiple sources in ItemsControl

### DISCUSSION THREAD

**Discussion by kosmakoff** [AvaloniaUI/Avalonia#14710] [Q&A] [answered]
I'm looking for a way to have multiple list sources (individual observable collections) to be rendered in `ItemsControl`, so that all items from all individual sources are shown as if one merged source was provided.

For context: I use `Canvas` as an Items Panel template, and individual sources are different types of items, like horizontal rulers, vertical rulers, etc, which I would rather have in separate lists as it is more convenient to modify them this way.

I tried to implement my own composite observable collection with signature as follows:
```c#
public class CompositeObservableCollection<TItem> :
    IReadOnlyList<TItem>,
    INotifyCollectionChanged,
    INotifyPropertyChanged
```

Alas, when I bind such source to `ItemsSource` property of `ItemsControl` it simply doesn't show anything. Debug shows that individual elements are not even being requested, `Count` property is not requested either.

Does `ItemsSource` literally have to be `ObservableCollection<>` type or am I missing something? Also, if there is any other way to show multiple sources in same list please let me know.

**Comment by BAndysc** [comment, marked as answer]
https://github.com/AvaloniaUI/Avalonia/blob/release/11.0.9/src/Avalonia.Controls/ItemsSourceView.cs#L267 I am surprised it doesn't throw an exception for you now. Implement IList and it should work

**Reply by kosmakoff** [reply]
Thanks a lot!!! Btw, is there a way to merge lists visually in any other way? Cause it feels like I am overcomplicating things.

**Reply by BAndysc** [reply]
Unfortunately I don't know of any other way (but maybe there is one, dunno)

**Reply by kosmakoff** [reply]
Unfortunately, seems like that's not it. Implementing `IList` didn't help.

**Reply by BAndysc** [reply]
You are doing something wrong, this works just fine
```
public class CustomCollection : IList, INotifyCollectionChanged
{
    private List<object> elements = new();

    public IEnumerator GetEnumerator()
    {
        return ((IEnumerable)elements).GetEnumerator();
    }

    public void CopyTo(Array array, int index)
    {
        ((ICollection)elements).CopyTo(array, index);
    }

    public int Count => elements.Count;

    public bool IsSynchronized => ((ICollection)elements).IsSynchronized;

    public object SyncRoot => ((ICollection)elements).SyncRoot;

    public int Add(object? value)
    {
        var index = ((IList)elements).Add(value);
        CollectionChanged?.Invoke(this, new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Add, value, index));
        return index;
    }

    public void Clear()
    {
        elements.Clear();
        CollectionChanged?.Invoke(this, new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
    }

    public bool Contains(object? value)
    {
        return ((IList)elements).Contains(value);
    }

    public int IndexOf(object? value)
    {
        return ((IList)elements).IndexOf(value);
    }

    public void Insert(int index, object? value)
    {
        ((IList)elements).Insert(index, value);
        CollectionChanged?.Invoke(this, new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Add, value, index));
    }

    public void Remove(object? value)
    {
        ((IList)elements).Remove(value);
        CollectionChanged?.Invoke(this, new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Remove, value));
    }

    public void RemoveAt(int index)
    {
        var element = elements[index];
        elements.RemoveAt(index);
        CollectionChanged?.Invoke(this, new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Remove, element, index));
    }

    public bool IsFixedSize => ((IList)elements).IsFixedSize;

    public bool IsReadOnly => ((IList)elements).IsReadOnly;

    public object? this[int index]
    {
        get => ((IList)elements)[index];
        set => ((IList)elements)[index] = value;
    }

    public event NotifyCollectionChangedEventHandler? CollectionChanged;
}
```

(probably no exception I have mentioned before is also a symptom of your problem). How are you binding the collection? Are you able to show some code snippets?

**Reply by kosmakoff** [reply]
Silly me. I had complex initialization logic for this collection, therefore it was in constructor. But I put the initialization after `InitializeComponent` is called, therefore it was simply null. Works now. Many thanks.
